### PR TITLE
Incorporating height profile into GrainStats

### LINKED ISSUE #748: Add horizontal grain profiles
**Is your feature request related to a problem? Please describe.**
It has been requested that topostats have a feature similar to Gwyddion where the horizontal (in the image) height profile of a grain would be taken. In other words, a horizontal line is placed over the molecule, and the pixels that form the line have their heights measured to form a profile for that molecule.

**Describe the solution you'd like**
During grainstats, a horizontal, one pixel thick line of pixels should be formed, with the length sufficient to cover the entire molecule from left to right, and the line intersecting the centroid of the molecule. The heights for those pixels should be obtained and saved as a line plot. Probably only saved with image_set all.

While it does seem arbitrary to just take one line through the centre, and we will miss a lot of information, this is is not an issue, as this line profile is just to get an idea of the height profile of the molecule.

This is a heavily requested feature and should be quite high priority.
It should be relatively trivial to implement. (I think?)

### PR DESCRIPTION
**Pull request by ns-rse**
Closes #748 

After the protracted wrestling I had of how to get minimum feret distances out (see #755) I've decided to finish this work off now as I'm unlikely to look at it again until September and had been tinkering with it over the last week and it is more efficient to finish it off now than struggle to remember where I was in two months time.

This PR includes two commits, importantly one fixes a problem with the `topostats.measure.height_profiles.interpolate_height_profiles()` which was introduced in #859 (details on that below). 

## Height Profiles

The option `extract_height_profile` has been introduced to the `grainstats` section of `default_config.yaml`, when `true` this will trigger extraction of height profiles along the maximum feret. These are aggregated across all "grains" and images and written to a JSON file (`height_profiles.json`) adjacent to `all_statistics.csv` in the specified output directory.

JSON is used because CSV is not particularly conducive to this data structure as profiles are taken across cropped images which differ in size.

This necessitated adding a dependency for `numpyencoder` which correctly and safely converts Numpy `dtypes` to a format that `json.dump()` can write.

I have a suspicion that at some point it will be desirable to generate line plots of these which is possible but for now users will have to extract the data align and plot these themselves (if anyone does this and wants to share or contribute their solution to TopoStats that would be very much appreciated).

## Fixing `interpolate_height_profiles()` #859

In #859 I had mistakenly used `np.sort()` on the feret co-ordinates. Two of tests developed there passed, one I opted to skip (The maximum feret was almost horizontal and this should have raised flags as it was indicative of another problem, namely that we should interpolate in such a manner as to get as many points as possible along the line of the maximum feret)

When it came to developing tests to incorporate height profile extraction into the work flow of `GrainStats()` it revealed that things were not as they should be as whilst the feret coordinates were correct point interpolation resulted in...

grain_with_sort01
grain_with_sort02
grain_with_sort03

Clearly not what we wanted! This arose because [`np.interp()`](https://numpy.org/doc/stable/reference/generated/numpy.interp.html) expects "_monotonically **increasing** sample points_" and in some of the above the feret coordinates were returned with the highest value first and were therefore **decreasing**.

### Solution

In order to get the interpolation correct after identifying these problems two changes were made.

1. Assess which is the largest distance between the feret co-ordinates on the x and y axis and the largest has points linearly interpolated first.
2. Order the feret points in ascending order (depending on whether it is the x or y axis that is the largest distance as this is the one along which interpolation will occur).

The points are now correctly interpolated regardless of the order in which the feret coordinates are returned and we get the greatest detail by doing so along the longest x/y axis first.

grain_correct01
grain_correct02
grain_correct03

As a bonus the test that was previously skipped is now looking as expected and has been reinstated.

### CODE REVIEW

**Comment by ns-rse** [comment]
Ready for review as I figured out how to update the HDF5 (see #873 for details and future reference).

In fixing this I found there was a precision error in the heights on OSX compared to Linux (where the tests were developed) to which end I split the regression of test of heights out to their own functions and skip them if the reported `platform.platform()[:5] == "macOS"`. This adds a slight increased overhead as in essence two of the tests are run twice but assessed differently but I couldn't see an option for format float's when using `json.dump()`.

I also noticed a missing f-string in one of the `plottingfuncs.py` `LOGGER.info()` so updated that.

**Review by SylviaWhittle — APPROVED** [review]
This works for me, and I am able to extract the height profiles from the HDF5 file and plot them. 

Thank you very much for your work on this 🎉

**Comment by SylviaWhittle** [comment]
Here is how I plotted the heights in case anyone reads this 
image

File: tests/measure/test_height_profiles.py [hunk 1]
```diff
@@ -398,4 +509,5 @@ def test_interpolate_height_profile_images(img: dict, interpolate_conf: dict, ta
     """Test interpolate_height_profile() with more realistic images."""
     _img = request.getfixturevalue(img)
     interpolated_heights = height_profiles.interpolate_height_profile(_img["img"], _img["skeleton"], **interpolate_conf)
+    print(f"{interpolated_heights=}")
```
[the remark this answers is not in the document]
Comment (ns-rse) [reply tests/measure/test_height_profiles.py]: Nope, shouldn't be there, not sure why `pylint`/`ruff` didn't flag this as an error as we have to explicitly disable it in the regression tests where `print()` is used to write the output to file.

Anyway, thanks for spotting this, I've removed it.

**Comment by ns-rse** [comment]
> Here is how I plotted the heights in case anyone reads this image

It should be possible to iteratie over the dictionary directly using `.items()`. Following is untested but should achieve the same thing.

```python
from pathlib import Path
from loguru import logger
import numpy as np
import h5py
import matplotlib.pyplot as plt

FILE_PATH = Path("./minicircle.topostats")
with h5py.File(FILE_PATH, "r") as f:
    # print keys
    logger.info(f.keys())
    height_profiles = f["height_profiles"]["above"]

fig, ax = plt.subplots(figsize=(10,5))
for key, profile in height_profiles.items()
    ax.plot(profile, label=key)
ax.set_xlabel("Pixel")
ax.set_ylabel("Height (nm)")
ax.set_title("Height Profiles")
plt.legend()
plt.show()
```

**Review by SylviaWhittle — APPROVED** [review]
🎉

**Pull request merged**